src/core/lib/iomgr/ev_epoll1_linux.cc

```diff
@@ -332,7 +341,15 @@ static void fd_shutdown_internal(grpc_fd* fd, grpc_error* why,
                                  bool releasing_fd) {
   if (fd->read_closure->SetShutdown(GRPC_ERROR_REF(why))) {
     if (!releasing_fd) {
-      shutdown(fd->fd, SHUT_RDWR);
+#ifdef GRPC_POSIX_FORK_ALLOW_PTHREAD_ATFORK
+      if (fd->fork_epoch < grpc_core::Fork::GetForkEpoch()) {
+        close(fd->fd);
```

Comment:
I'm not sure this is safe to do.

See comment below, ```fd_shutdown``` might be called multiple times.

Reply:
I think the read_closure->SetShutdown() call ensures this branch is only reached one time. Otherwise, this is unsafe (the fd number may have been reused by a new endpoint) - but this would also be unsafe for the code's as-is invocation of shutdown, which could otherwise shutdown the wrong endpoint on a second call to this function.

Reply:
Makes sense.